text/0000-custom-components.md

```diff
@@ -0,0 +1,1078 @@
+
+The *capabilities* property specifies the optional features required by this
+component. This will be described in more detail below.
+
+Finally, there is an optional *metadata* property which component authors can
+use to store arbitrary data. For example, it may include a *class* property to
+specify which component class to use. The *metadata* property is ignored by
+Ember but can be used by the `ComponentManager` to perform custom logic.
+
+## `ComponentManager`
+
+Whereas a `ComponentDefinition` describe the static property of the component,
+a `ComponentManager` controls its runtime properties.
+
+A basic `ComponentManager` satisfies the following interface:
+
+```typescript
+interface ComponentManager<T> {
+  create(definition: ComponentDefinition, args: ComponentArguments): T;
+  getContext(instance: T): any;
+  update(instance: T, args: ComponentArguments): void;
+}
+
+interface ComponentArguments {
+  positional: any[];
+  named: Object;
+}
+```
+
+> **Open Question**: Should we require `ComponentManager` to inherit from a
+provided super class (`Ember.ComponentManager.extend({ ... }`)?
+
+When Ember is about to render a component, Ember will lookup its component
+manager (as described above) and call its *create* method with the component
+defition and the *component arguments*.
+
+The *component arguments* object is a snapshot of the arguments pased into the
+component. It has a *positional* and *named* property which corresponds to an
+array and object (dictionary) of the current argument values.
+
+For example, for the following invocation:
+
+```hbs
+{{blog-post (titleize post.title) post.body author=post.author excerpt=true}}
+```
+
+Glimmer will look up the `blog-post` definition and call *create* on its
+manager with the following `ComponentArguments`:
+
+```javascript
+{
+  positional: [
+    "Rails Is Omakase",
+    "Lorem ipsum dolor sit amet, consectetur adipiscing elit..."
+  ],
+  named: {
+    "author": #<User name="David Heinemeier Hansson", ...>,
+    "excerpt": true
+  }
+}
+```
+
+The component manager *must not* mutate the *component arguments* object (and
+the inner *positional* and *named* object/array) directly as they might be
+pooled or reused by the system.
+
+> **Note**: We should probably freeze them in debug mode.
+
+Based on the information in the definition and the *component arguments*, the
+manager should return a *component instance* from the *create* method. From
+Ember's perspective, this could be any arbitrary value – it is only used for
+the component manager's internal book-keeping. In practice, this value should
+store enough information to represent the internal state of the component. For
+these reasons, it is often referred to as the "opaque state bucket".
+
+At a later time, before Ember is ready to render the template, the component
+manager's *getContext* method will be called. It will receive the *component
+instance* and return the context object for the template. The context binds
```

Comment:
> The component instance is for your own internal use, so it could be anything. It's whatever you returned from create.

Not sure I understand the reasoning behind returning _anything_.